### ISSUE TITLE
GetDataSetDirectoryAsync got System.AccessViolationException

### ISSUE THREAD

**Issue by Tivaty** [mz-automation/libiec61850#498]
Hello,

Evinronment: C#
Lib version 1.5.3

When I set MaxPduSize to 1000 app crash happens. 
The reason I need to set it to 1000 to GetLogicalDeviceVariablesAsync without problem of timeout. When the value of MaxPduSize is larger than 1000 IED timeout error happens frequently.

```
IedConnection con = new IedConnection();
con.MaxPduSize = 1000; // GetDataSetDirectoryAsync and GetDataSetDirectory Failed
```

`System.AccessViolationException
  HResult=0x80004003
  Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>
`

<img width="780" alt="Untitled" src="https://github.com/mz-automation/libiec61850/assets/79079033/feab60f9-fd2a-412c-8d13-01666076403a">

**Comment by mzillgith** [comment]
In what function does the application crash?
Do you have a stack trace?

**Comment by Tivaty** [comment]
> In what function does the application crash? Do you have a stack trace?

Hi,
This is C# library so there is no stacktrace, only error message.

App crashed at "con.GetDataSetDirectoryAsync"

```
string dsRef = dataSetRef.Replace('$', '.');
Debug.WriteLine(dsRef);
waitForCallBack.Reset();
con.GetDataSetDirectoryAsync(dsRef, delegate (uint invokeId, object parameter, IedClientError err, List<string> dataSetDirectory, bool isDeletable)
{
	if (dataSetDirectory != null)
	{
		foreach (var doName in dataSetDirectory)
		{
			Debug.WriteLine("\t" + doName);
		}
	}
	else
	{
		Debug.WriteLine("Get dataset directory error: " + err.ToString());
	}
	waitForCallBack.Set();
}, null);
waitForCallBack.WaitOne(); 
```

**Comment by Tivaty** [comment]
Please note that, when con.MaxPduSize does not set. (default value 65000) this function works well. 
And I have 2 dataset LLN0$PROT1 have much more variable than LLN0$PROT2.

**Comment by mzillgith** [comment]
I found an issue with the .NET wrapper when the server doesn't send a response but an error message instead. Could this be the case here? I guess the dataset is too big to fit into a the small MMS PDU size and the server is sending an error message instead.

**Comment by Tivaty** [comment]
> I found an issue with the .NET wrapper when the server doesn't send a response but an error message instead. Could this be the case here? I guess the dataset is too big to fit into a the small MMS PDU size and the server is sending an error message instead.

You are correct. And the issue I want to point out is the dataset is too big to fit into a the small MMS PDU size. 
In the GetNameListHandler there is a bool moreFollow. But with GetDataSetDirectoryHandler there is only isDeletable.

the reason PDU size is 1000 is for GetNameListHandler work well, when PDU larger than 1000 I got many IED timeout with my test.

**Comment by mzillgith** [comment]
There is no way to send a larger dataset when the MMS size is too small. Only the getnamelist MMS service supports splitting the response in multiple MMS messages.

But the issue (crash) should be resolved with this fix: https://github.com/mz-automation/libiec61850/commit/681d1b0e05e83bf6d45e42b66c9c780df3f40415

**Comment by Tivaty** [comment]
> There is no way to send a larger dataset when the MMS size is too small. Only the getnamelist MMS service supports splitting the response in multiple MMS messages.
> 
> But the issue (crash) should be resolved with this fix: [681d1b0](https://github.com/mz-automation/libiec61850/commit/681d1b0e05e83bf6d45e42b66c9c780df3f40415)

Okay. Thank youuuu. Let me try again with the latest source code
